core/kernel.go

```diff
@@ -73,300 +65,305 @@ const (
-			Enabled: rpcConfig.Info.Enabled,
-			Launch: func() (process.Process, error) {
-				server, err := rpcinfo.StartServer(kern.Service, "/websocket", rpcConfig.Info.ListenAddress, kern.Logger)
-				if err != nil {
-					return nil, err
-				}
-				return server, nil
-			},
-		},
-		{
-			Name:    "RPC/metrics",
-			Enabled: rpcConfig.Metrics.Enabled,
-			Launch: func() (process.Process, error) {
-				server, err := metrics.StartServer(kern.Service, rpcConfig.Metrics.MetricsPath,
-					rpcConfig.Metrics.ListenAddress, rpcConfig.Metrics.BlockSampleSize, kern.Logger)
-				if err != nil {
-					return nil, err
-				}
-				return server, nil
-			},
-		},
-		{
-			Name:    "RPC/GRPC",
-			Enabled: rpcConfig.GRPC.Enabled,
-			Launch: func() (process.Process, error) {
-				listen, err := net.Listen("tcp", rpcConfig.GRPC.ListenAddress)
-				if err != nil {
-					return nil, err
-				}
 
-				grpcServer := rpc.NewGRPCServer(kern.Logger)
-				var ks *keys.KeyStore
-				if keyStore != nil {
-					ks = keyStore
-				}
-
-				if keyConfig.GRPCServiceEnabled {
-					if keyStore == nil {
-						ks = keys.NewKeyStore(keyConfig.KeysDirectory, keyConfig.AllowBadFilePermissions)
-					}
-					keys.RegisterKeysServer(grpcServer, ks)
-				}
+	tmGenesisDoc := tendermint.DeriveGenesisDoc(&genesisDoc)
+	heightValuer := log.Valuer(func() interface{} { return kern.Blockchain.LastBlockHeight() })
+	tmLogger := kern.Logger.With(structure.CallerKey, log.Caller(LoggingCallerDepth+1))
+	tmLogger = tmLogger.With("height", heightValuer)
+	kern.Node, err = tendermint.NewNode(tmConf.TendermintConfig(), privVal, tmGenesisDoc, app, metricsProvider, nodeKey, tmLogger)
+	return err
+}
 
-				rpcquery.RegisterQueryServer(grpcServer, rpcquery.NewQueryServer(kern.State, nameRegState, proposalRegState,
-					kern.Blockchain, kern.State, nodeView, kern.Logger))
+func (kern *Kernel) LoadTransactor() (err error) {
+	nodeView, err := kern.GetNodeView()
+	if err != nil {
+		return err
+	}
 
-				rpctransact.RegisterTransactServer(grpcServer, rpctransact.NewTransactServer(kern.Transactor, txCodec))
+	accountState := kern.State
+	nameRegState := kern.State
+	kern.Service = rpc.NewService(accountState, nameRegState, kern.Blockchain, kern.State, nodeView, kern.Logger)
 
-				rpcevents.RegisterExecutionEventsServer(grpcServer, rpcevents.NewExecutionEventsServer(kern.State,
-					kern.Emitter, kern.Blockchain, kern.Logger))
+	if nodeView == nil {
+		checkTx := func(tx tmTypes.Tx, cb func(*abciTypes.Response)) error {
+			ctx := abci.TxExecutor("CheckTx", kern.exeChecker, kern.txCodec, kern.Logger.WithScope("CheckTx"))(tx)
+			dtx := abci.TxExecutor("DeliverTx", kern.exeCommitter, kern.txCodec, kern.Logger.WithScope("DeliverTx"))(tx)
+			appHash, err := kern.exeCommitter.Commit(nil)
+			if err != nil {
+				return err
+			}
+			kern.Blockchain.CommitBlock(time.Now(), nil, appHash)
+
+			cb(abciTypes.ToResponseCheckTx(ctx))
+			cb(abciTypes.ToResponseDeliverTx(abciTypes.ResponseDeliverTx{
```

Comment:
I'm just wondering if it worth introducing a little type that wraps this up:

```
type TxExecutor {
	checker execution.BatchExecutor
	deliverer execution.BatchExecutor
	txDecoder txs.Decoder
	logger *logging.Logger
}

func (txx TxExecutor) CheckTx(txBytes []byte) abciTypes.ResponseCheckTx {
	return txx.execute("CheckTx", txx.checker, txBytes)
}

func (txx TxExecutor) DeliverTx(txBytes []byte) abciTypes.ResponseDeliverTx {
	ctr := txx.execute("DeliverTx", txx.deliverer, txBytes)
	return abciTypes.ResponseDeliverTx{
		Code:      ctr.Code,
		Log:       ctr.Log,
		Data:      ctr.Data,
		Tags:      ctr.Tags,
		GasUsed:   ctr.GasUsed,
		GasWanted: ctr.GasWanted,
		Info:      ctr.Info,
	}
}

func executor(name string, executor execution.BatchExecutor, txBytes []byte) {
...
}
```
Then we can just embed that in `App` and reuse in kernel

Reply:
I like that :)